src/d3a/models/area/redis_dispatcher/market_event_dispatcher.py

```diff
@@ -0,0 +1,122 @@
+import json
+from random import random
+from copy import deepcopy
+from d3a.events import MarketEvent
+from d3a.models.area.redis_dispatcher import RedisEventDispatcherBase
+from d3a.models.market.market_structures import trade_from_JSON_string, offer_from_JSON_string
+
+from threading import Event
+
+
+class RedisMarketEventDispatcher(RedisEventDispatcherBase):
+    def __init__(self, area, root_dispatcher, redis):
+        super().__init__(area, root_dispatcher, redis)
+        self.str_market_events = [event.name.lower() for event in MarketEvent]
+        self.active_trade = False
+        self.deferred_events = []
+        self.events_to_wait = {
+            MarketEvent.OFFER: Event(),
+            MarketEvent.OFFER_CHANGED: Event(),
+            MarketEvent.TRADE: Event(),
+            MarketEvent.OFFER_DELETED: Event()
+        }
+
+    def event_channel_name(self):
+        return f"{self.area.slug}/market_event"
+
+    def event_response_channel_name(self):
+        return f"{self.area.slug}/market_event_response"
+
+    def event_listener_redis(self, payload):
+        event_type, kwargs = self.parse_market_event_from_event_payload(payload)
+
+        # If this is a trade-related event, it needs to be executed immediately, all other
+        # events need to be stored for deferred execution
+        if self.active_trade and event_type not in self._trade_related_events:
+            self.store_deferred_events_during_active_trade(payload)
+        # If there is no active trade but we still have deferred events, we need to handle these.
+        elif not self.active_trade and len(self.deferred_events) > 0:
+            self.store_deferred_events_during_active_trade(payload)
+            self.run_deferred_events()
+        # No deferred event, handle directly the event
+        else:
+            # Consume already deferred events before handling the new one
+            from concurrent.futures import ThreadPoolExecutor
+            executor = ThreadPoolExecutor(max_workers=1)
+
+            def executor_func():
+                self.root_dispatcher.event_listener(event_type=event_type, **kwargs)
+                self.publish_response(event_type)
+
+            executor.submit(executor_func)
```

Comment:
It seams that this could be refactored into a separate function (in e.g. RedisEventDispatcherBase) ´ and only creating executor_func here. 
If not, please move the lazy import to the top. Thanks!

Reply:
I think it has already been adapted to what you suggested, thanks!